File: python/longrep.py


'''
LONGEST SUBSTRING WITHOUT REPEATING CHARACTERS
substring has to be continguous


basically have 1 substr, 
if theres a duplicate remove it and all chars on the left from the substr

make the substr longer

SLIDING WINDOW TECHNIQUE <- very common
O(n) space time complexity
O(n) memory complexity

maintain longest substring

    
'''

class Solution:
    def sub_str_len(self, s:str) -> int:
        char_set = set()
        result = 0


        #sliding window - > left and right ptr
        left = 0 # index of leftmox char, starts at 0 and increases by 1
        for right in range(len(s)):
            while s[right] in char_set: #while the duplicate is within the character set
                char_set.remove(s[left]) # removes every left and duplicate
                left +=1
            char_set.add(s[right])
            result = max(result, right-left +1) #checks if the current is bigger than the result
            # if so, assigns it to result
        return result


print(Solution.sub_str_len(Solution, 'abcdabcdc'))

File: cpp/longrep.cpp
#include <iostream>
#include <vector>
#include <string>
#include <unordered_set>

/*
LONGEST SUBSTRING WITHOUT REPEATING CHARACTERS

SLIDING WINDOW TECHNIQUE
with a set

clang++ longrep.cpp -o linout/longrep
linout/longrep

i and j pointer poitns to begin ning of string
and output int 

init a set that keeps track of unique characters
if a character is not in a set, add p to it, and compute the length of window
length will always be i-j+1 , +1 because i and j are indeces

move i , compute new length 
if s[i] is in a set, move j to i, remove all values in the set, and add the value s[i]

output will be maximum of output and current window
*/

class Solution {
public:
    static int lengthOfLongestSubstring(const std::string& s) {
        if(!s.length()) return 0;

        int i=0, j=0, max = 0;
        //key, value
        std::unordered_set<char> set;
        for(int i=0; i<s.length(); ++i){
            char curr = s.at(i);
            while(set.contains(curr)){
                set.erase(s.at(j));
                ++j;
            }
            set.insert(curr);
            max = std::max(max, i-j+1);
            }
        return max;
        }
};

int main(){

    std::cout<< Solution::lengthOfLongestSubstring("abcabcdef");


}